### PR TITLE
Upgrade WABT to 1.0.19

### PR DESCRIPTION
**Pull request by chfast**
For reference, here's PR intoducing new API: https://github.com/WebAssembly/wabt/pull/1330

### CODE REVIEW

File: test/utils/wabt_engine.cpp [hunk 1]
```diff
@@ -37,123 +49,126 @@ std::unique_ptr<WasmEngine> create_wabt_engine()
 
 bool WabtEngine::parse(bytes_view input) const
 {
-    wabt::interp::Environment env;
-
-    wabt::interp::HostModule* hostModule = env.AppendHostModule("env");
-    assert(hostModule != nullptr);
-
-    hostModule->AppendFuncExport("adler32", {{wabt::Type::I32, wabt::Type::I32}, {wabt::Type::I32}},
```
Comment (axic) [inline test/utils/wabt_engine.cpp:45]: Nice, it doesn't require this for parsing anymore?

Comment (gumb0) [reply test/utils/wabt_engine.cpp:45]: Yes, parsing is now a normal separate step.

File: test/utils/wabt_engine.cpp [hunk 2]
```diff
@@ -37,123 +49,126 @@ std::unique_ptr<WasmEngine> create_wabt_engine()
...
-        wasm_binary.size(), wabt::ReadBinaryOptions{}, &errors, &m_module);
-    if (result != wabt::Result::Ok)
-        return false;
-    assert(m_module != nullptr);
+    wabt::interp::Trap::Ptr trap;
+    m_instance =
+        wabt::interp::Instance::Instantiate(m_store, m_module.ref(), {host_func->self()}, &trap);
 
-    // Run start function (this will return ok if no start function is present)
-    const wabt::interp::ExecResult r = m_executor.RunStartFunction(m_module);
```
Comment (axic) [inline test/utils/wabt_engine.cpp:89]: Is this now executed properly by instantiate?

File: test/utils/wabt_engine.cpp [hunk 3]
```diff
@@ -37,123 +49,126 @@ std::unique_ptr<WasmEngine> create_wabt_engine()
...
-    return r.result == wabt::interp::Result::Ok;
+    return m_instance && !trap;
 }
 
 bool WabtEngine::init_memory(bytes_view memory)
 {
-    if (m_env.GetMemoryCount() == 0)
-        return false;
+    if (!m_instance || m_instance->memories().empty())
+        return {};
```
Comment (axic) [inline test/utils/wabt_engine.cpp:99]: Why not keep false here?

File: test/utils/wabt_engine.cpp (hunk 2 above)
Comment (gumb0) [reply test/utils/wabt_engine.cpp:89]: Yes, looks like so, our unit test for this passes.

File: test/utils/wabt_engine.cpp (hunk 3 above)
Comment (gumb0) [reply test/utils/wabt_engine.cpp:99]: typo, thanks

File: test/utils/wabt_engine.cpp [hunk 4]
```diff
@@ -37,123 +49,128 @@ std::unique_ptr<WasmEngine> create_wabt_engine()
...
+
+    return WasmEngine::FuncRef{it_export->index};
 }
 
 WasmEngine::Result WabtEngine::execute(
     WasmEngine::FuncRef func_ref, const std::vector<uint64_t>& args)
 {
-    const auto* e = reinterpret_cast<const wabt::interp::Export*>(func_ref);
+    assert(m_instance);
```
Comment (axic) [inline test/utils/wabt_engine.cpp:137]: Why assert here and not in init_memory/get_memory? I think we have the "rules" described in `wasm_engine.h`, I can't remember myself.

Comment (gumb0) [reply test/utils/wabt_engine.cpp:138]: `wasm_engine.hpp` says for all three`Requires instantiate().`, though it seems ambiguous whether it means "assumes instantiate was called" or "fails when instantiate was not called".

But I see that here it's inconsitent now, I guess I'll change it to asserts in all three, `fizzy_engine` also assumes  `m_instance` is non-null there.

File: test/utils/wabt_engine.cpp [hunk 5]
```diff
@@ -8,17 +8,29 @@
...
 #include <test/utils/wasm_engine.hpp>
+#include <algorithm>
 #include <cassert>
 
 namespace fizzy::test
 {
-class WabtEngine final : public WasmEngine
+namespace
+{
+const wabt::Features& get_features() noexcept
```
Comment (chfast) [inline test/utils/wabt_engine.cpp]: This is cheap object so can be returned by value. I guess it is taken by reference by other functions?

File: test/utils/wabt_engine.cpp [hunk 6]
```diff
@@ -8,17 +8,29 @@
...
 {
-    wabt::interp::Environment m_env;
-    wabt::interp::DefinedModule* m_module{nullptr};
+    static wabt::Features features;
```
Comment (chfast) [inline test/utils/wabt_engine.cpp]: Construct with lambda to disable flags once. Constexpr if possible.

File: test/utils/wabt_engine.cpp [hunk 7]
```diff
@@ -8,17 +8,23 @@
...
 class WabtEngine final : public WasmEngine
 {
-    wabt::interp::Environment m_env;
-    wabt::interp::DefinedModule* m_module{nullptr};
-
-    // WABT Executor/Thread with default options.
-    wabt::interp::Executor m_executor{&m_env};
+    // mutable because getting RefPtr from Ref requires non-const Store
+    mutable wabt::interp::Store m_store{[] {
+        static wabt::Features features;
```
Comment (chfast) [inline test/utils/wabt_engine.cpp:20]:
```suggestion
        wabt::Features features;
```
The `static` objects are actually expensive as they must be thread-safe - there might be a mutex used.

File: test/utils/wabt_engine.cpp [hunk 8]
```diff
@@ -37,123 +43,131 @@ std::unique_ptr<WasmEngine> create_wabt_engine()
...
         else
-            value.i64 = args[i];
-        typed_args.push_back(wabt::interp::TypedValue{type, value});
+            value.Set(args[i]);
+        typed_args.emplace_back(value);
     }
-    wabt::interp::ExecResult r = m_executor.RunExport(e, typed_args);
 
-    if (r.result != wabt::interp::Result::Ok)
+    wabt::interp::Values results;
```
Comment (chfast) [inline test/utils/wabt_engine.cpp:153]: Having `results` and `result` variables is confusing.

File: test/utils/wabt_engine.cpp [hunk 9]
```diff
@@ -8,17 +8,23 @@
...
 {
 class WabtEngine final : public WasmEngine
 {
-    wabt::interp::Environment m_env;
-    wabt::interp::DefinedModule* m_module{nullptr};
-
-    // WABT Executor/Thread with default options.
-    wabt::interp::Executor m_executor{&m_env};
+    // mutable because getting RefPtr from Ref requires non-const Store
+    mutable wabt::interp::Store m_store{[]() noexcept {
```
Comment (gumb0) [inline test/utils/wabt_engine.cpp:19]: Lol when adding `constexpr` here clang-format gets crazy
```suggestion
    mutable wabt::interp::Store m_store{[]() constexpr noexcept {
```

Comment (chfast) [reply test/utils/wabt_engine.cpp:19]: Good to report bug.

File: test/utils/wabt_engine.cpp (hunk 7 above)
Comment (gumb0) [reply test/utils/wabt_engine.cpp]: I changed this, but what mutex?

File: test/utils/wabt_engine.cpp (hunk 9 above)
Comment (gumb0) [reply test/utils/wabt_engine.cpp]: Additional pair of parentheses helped.

File: test/utils/wabt_engine.cpp [hunk 10]
```diff
@@ -37,123 +43,132 @@ std::unique_ptr<WasmEngine> create_wabt_engine()
...
-            const auto offset = args[0].value.i32;
-            const auto length = args[1].value.i32;
+    wabt::Errors errors;
+    wabt::interp::ModuleDesc module_desc;
+    const wabt::Result result = wabt::interp::ReadBinaryInterp(
+        wasm_binary.data(), wasm_binary.size(), wabt::ReadBinaryOptions{}, &errors, &module_desc);
+    if (result != wabt::Result::Ok)
+        return false;
+
+    m_module = wabt::interp::Module::New(m_store, module_desc);
```
Comment (axic) [inline test/utils/wabt_engine.cpp:62]: Is it a valid error condition that this returns `nullptr`? If yes add a condition, if not just add `assert(m_module != nullptr)`.

Comment (axic) [reply test/utils/wabt_engine.cpp:62]: Oh it is not a ptr, I thought it is given it is called `::Ptr`.

Comment (gumb0) [reply test/utils/wabt_engine.cpp:62]: It is a smart pointer, and it returns always non-null, will add an assertion.

File: test/utils/wabt_engine.cpp [hunk 11]
```diff
@@ -37,123 +43,140 @@ std::unique_ptr<WasmEngine> create_wabt_engine()
...
+    assert(m_module);
+
+    auto host_func = wabt::interp::HostFunc::New(m_store,
```
Comment (axic) [inline test/utils/wabt_engine.cpp:65]: Is this also a smart pointer? Perhaps add an assertion here too for `host_func`.

File: test/utils/wabt_engine.cpp [hunk 12]
```diff
@@ -37,123 +43,140 @@ std::unique_ptr<WasmEngine> create_wabt_engine()
...
 }
 
 std::optional<WasmEngine::FuncRef> WabtEngine::find_function(
     std::string_view name, std::string_view) const
 {
-    const wabt::interp::Export* e = m_module->GetExport({name.data(), name.size()});
-    if (e != nullptr && e->kind == wabt::ExternalKind::Func)
-        return reinterpret_cast<WasmEngine::FuncRef>(e);
-    return {};
+    if (!m_module)
```
Comment (axic) [inline test/utils/wabt_engine.cpp:124]: I think this should be an assertion instead.

**Comment by axic** [comment]
Can you also squash the commits? I think this is ready now, isn't it?

**Review by axic — APPROVED** [review]
Looks good to me, we should have a benchmarking comparing this to 1.0.12 though.

**Comment by axic** [comment]
@chfast waiting for your review

**Comment by chfast** [comment]
```
 wabt/parse/blake2b_mean                                          -0.0325         -0.0325            72            70            72            70                       
 wabt/instantiate/blake2b_mean                                    +0.0876         +0.0876            78            85            78            85                       
 wabt/execute/blake2b/512_bytes_rounds_1_mean                     +2.0244         +2.0244           154           466           154           466                       
 wabt/execute/blake2b/512_bytes_rounds_16_mean                    +2.0314         +2.0314          2356          7141          2356          7141                       
 wabt/parse/ecpairing_mean                                        +0.0098         +0.0098          4115          4156          4115          4156                       
 wabt/instantiate/ecpairing_mean                                  +0.0455         +0.0455          4148          4336          4147          4336                       
 wabt/execute/ecpairing/onepoint_mean                             +2.0233         +2.0233        720502       2178284        720507       2178307                       
 wabt/parse/keccak256_mean                                        -0.0074         -0.0074           120           119           120           119                       
 wabt/instantiate/keccak256_mean                                  +0.1054         +0.1054           123           136           123           136                       
 wabt/execute/keccak256/512_bytes_rounds_1_mean                   +1.5514         +1.5514           173           441           173           441                       
 wabt/execute/keccak256/512_bytes_rounds_16_mean                  +1.5498         +1.5498          2543          6485          2543          6485                       
 wabt/parse/memset_mean                                           -0.1456         -0.1455            19            16            19            16                       
 wabt/instantiate/memset_mean                                     +0.1560         +0.1560            23            27            23            27
 wabt/execute/memset/256_bytes_mean                               +2.2922         +2.2921            11            38            11            38
 wabt/execute/memset/60000_bytes_mean                             +2.2041         +2.2041          2521          8077          2521          8077
 wabt/parse/mul256_opt0_mean                                      -0.1522         -0.1522            24            20            24            20
 wabt/instantiate/mul256_opt0_mean                                +0.1650         +0.1650            27            32            27            32
 wabt/execute/mul256_opt0/input0_mean                             +2.5594         +2.5594            44           158            44           158
 wabt/execute/mul256_opt0/input1_mean                             +2.5540         +2.5540            45           158            45           158
 wabt/parse/ramanujan_pi_mean                                     -0.0393         -0.0393            65            62            65            62
 wabt/instantiate/ramanujan_pi_mean                               +0.1558         +0.1558            68            78            68            78
 wabt/execute/ramanujan_pi/33_runs_mean                           +1.8711         +1.8711           237           679           237           679
 wabt/parse/sha1_mean                                             -0.0246         -0.0246           113           110           113           110
 wabt/instantiate/sha1_mean                                       +0.0554         +0.0554           116           123           116           123
 wabt/execute/sha1/512_bytes_rounds_1_mean                        +2.2715         +2.2715           174           568           174           568
 wabt/execute/sha1/512_bytes_rounds_16_mean                       +2.2881         +2.2881          2415          7940          2415          7940
 wabt/parse/sha256_mean                                           -0.0246         -0.0246           198           193           198           193
 wabt/instantiate/sha256_mean                                     +0.0445         +0.0445           202           211           202           211
 wabt/execute/sha256/512_bytes_rounds_1_mean                      +1.9267         +1.9267           188           551           188           551
 wabt/execute/sha256/512_bytes_rounds_16_mean                     +1.9418         +1.9418          2599          7647          2599          7647
 wabt/parse/taylor_pi_mean                                        -0.3610         -0.3610             9             6             9             6
 wabt/instantiate/taylor_pi_mean                                  +0.2191         +0.2191            14            17            14            17
 wabt/execute/taylor_pi/pi_1000000_runs_mean                      +1.7793         +1.7793         70885        197008         70885        197010
 wabt/parse/micro/eli_interpreter_mean                            -0.2105         -0.2105            12             9            12             9
 wabt/instantiate/micro/eli_interpreter_mean                      +0.1291         +0.1291            16            19            16            19
 wabt/execute/micro/eli_interpreter/halt_mean                    +19.4798        +19.4796             0             3             0             3
 wabt/execute/micro/eli_interpreter/exec105_mean                +152.1850       +152.1849             0            19             0            19
 wabt/parse/micro/factorial_mean                                  -0.0874         -0.0874             2             2             2             2
 wabt/instantiate/micro/factorial_mean                            +0.5015         +0.5015             2             3             2             3
 wabt/execute/micro/factorial/10_mean                             +8.1540         +8.1538             0             3             0             3
 wabt/execute/micro/factorial/20_mean                             +5.6214         +5.6213             1             4             1             4
 wabt/parse/micro/fibonacci_mean                                  -0.0653         -0.0653             3             3             3             3
 wabt/instantiate/micro/fibonacci_mean                            +0.3994         +0.3994             3             4             3             4
 wabt/execute/micro/fibonacci/24_mean                             +2.1884         +2.1884          5812         18529          5812         18529
 wabt/parse/micro/host_adler32_mean                               -0.3049         -0.3049             5             3             5             3
 wabt/instantiate/micro/host_adler32_mean                         +0.1657         +0.1657             7             8             7             8
 wabt/execute/micro/host_adler32/1_mean                           +0.1405         +0.1405             2             3             2             3
 wabt/execute/micro/host_adler32/100_mean                         -0.9252         -0.9252           219            16           219            16
 wabt/execute/micro/host_adler32/1000_mean                        -0.9273         -0.9273          2184           159          2184           159
 wabt/parse/micro/spinner_mean                                    -0.1156         -0.1156             2             2             2             2
 wabt/instantiate/micro/spinner_mean                              +0.5635         +0.5635             2             3             2             3
 wabt/execute/micro/spinner/1_mean                               +44.1612        +44.1605             0             2             0             2
 wabt/execute/micro/spinner/1000_mean                             +1.8353         +1.8353            15            41            15            41
```

This does not look very impressive.

File: test/utils/wabt_engine.cpp [hunk 13]
```diff
@@ -8,17 +8,23 @@
...
 
 namespace fizzy::test
 {
 class WabtEngine final : public WasmEngine
 {
-    wabt::interp::Environment m_env;
-    wabt::interp::DefinedModule* m_module{nullptr};
-
-    // WABT Executor/Thread with default options.
-    wabt::interp::Executor m_executor{&m_env};
```
Comment (chfast) [inline test/utils/wabt_engine.cpp:21]: Previously the stack space was allocated here. Not allocating stack for every call had big impact on performance.

File: test/utils/wabt_engine.cpp [hunk 14]
```diff
@@ -8,17 +8,24 @@
...
+    mutable wabt::interp::Store m_store{([]() constexpr noexcept {
+        wabt::Features features;
+        features.disable_multi_value();
+        features.disable_sat_float_to_int();
+        features.disable_sign_extension();
+        return features;
+    })()};
+    wabt::interp::Module::Ptr m_module;
+    wabt::interp::Instance::Ptr m_instance;
+    wabt::interp::Thread::Ptr m_thread{wabt::interp::Thread::New(m_store, {})};
```
Comment (gumb0) [inline test/utils/wabt_engine.cpp:28]: Here are the options for `Thread`
https://github.com/WebAssembly/wabt/blob/21350996d5bbcd9cafbaa3a5687e6fb503da060c/src/interp/interp.h#L1025-L1032

Comment (chfast) [reply test/utils/wabt_engine.cpp:28]: Default values are fine provided we create it once.

Comment (axic) [reply test/utils/wabt_engine.cpp:28]: So `m_store` does not need to be mutable anymore, right?

**Comment by chfast** [comment]
I benchmarked the version with `Thread` preallocation, but it changes little. The `Thread` allocation saves ~2 µs so it affects mostly micro-benchmarks.

**Review by chfast — COMMENTED** [review]
Approved.

File: test/utils/wabt_engine.cpp (hunk 14 above)
Comment (gumb0) [reply test/utils/wabt_engine.cpp:28]: > So `m_store` does not need to be mutable anymore, right?

It still needs it for `get_memory` where it gets the (smart) pointer to memory, which can be only non-const.

**Comment by axic** [comment]
Make sure to squash the last commit :)

**Comment by gumb0** [comment]
I've got similar or worse results
```
 wabt/parse/blake2b_mean                                          -0.0195         -0.0202           149           146           149           146
 wabt/instantiate/blake2b_mean                                    +0.0806         +0.0807           158           171           158           171
 wabt/execute/blake2b/512_bytes_rounds_1_mean                     +3.0988         +3.1005           243           995           243           996
 wabt/execute/blake2b/512_bytes_rounds_16_mean                    +3.1907         +3.1922          3640         15254          3640         15260
 wabt/parse/ecpairing_mean                                        +0.0178         +0.0181          8392          8541          8392          8544
 wabt/instantiate/ecpairing_mean                                  +0.0387         +0.0390          8428          8754          8428          8756
 wabt/execute/ecpairing/onepoint_mean                             +3.0752         +3.0756       1114023       4539822       1114017       4540296
 wabt/parse/keccak256_mean                                        +0.0113         +0.0117           247           250           247           250
 wabt/instantiate/keccak256_mean                                  +0.1028         +0.1026           251           276           250           276
 wabt/execute/keccak256/512_bytes_rounds_1_mean                   +2.9105         +2.9107           237           926           237           926
 wabt/execute/keccak256/512_bytes_rounds_16_mean                  +2.9136         +2.9132          3480         13620          3479         13615
 wabt/parse/memset_mean                                           -0.1130         -0.1131            38            33            38            33
 wabt/instantiate/memset_mean                                     +0.2092         +0.2091            46            56            46            56
 wabt/execute/memset/256_bytes_mean                               +2.5077         +2.5075            21            72            21            72
 wabt/execute/memset/60000_bytes_mean                             +2.4606         +2.4603          4535         15695          4535         15691
 wabt/parse/mul256_opt0_mean                                      -0.1235         -0.1236            49            43            49            43
 wabt/instantiate/mul256_opt0_mean                                +0.0473         +0.0473            56            58            56            58
 wabt/execute/mul256_opt0/input0_mean                             +3.3520         +3.3518            79           344            79           344
 wabt/execute/mul256_opt0/input1_mean                             +3.3580         +3.3578            79           344            79           344
 wabt/parse/ramanujan_pi_mean                                     -0.0382         -0.0383           135           130           135           130
 wabt/instantiate/ramanujan_pi_mean                               +0.0804         +0.0804           141           153           141           153
 wabt/execute/ramanujan_pi/33_runs_mean                           +3.0658         +3.0658           325          1321           325          1321
 wabt/parse/sha1_mean                                             -0.0158         -0.0158           233           229           233           229
 wabt/instantiate/sha1_mean                                       +0.0663         +0.0663           237           252           237           252
 wabt/execute/sha1/512_bytes_rounds_1_mean                        +3.1111         +3.1111           269          1104           268          1104
 wabt/execute/sha1/512_bytes_rounds_16_mean                       +3.1225         +3.1225          3755         15478          3754         15477
 wabt/parse/sha256_mean                                           -0.0251         -0.0252           404           394           404           394
 wabt/instantiate/sha256_mean                                     +0.0205         +0.0205           411           419           411           419
 wabt/execute/sha256/512_bytes_rounds_1_mean                      +3.0870         +3.0869           252          1031           252          1031
 wabt/execute/sha256/512_bytes_rounds_16_mean                     +3.1396         +3.1395          3467         14354          3467         14353
 wabt/parse/taylor_pi_mean                                        -0.3111         -0.3111            18            13            18            13
 wabt/instantiate/taylor_pi_mean                                  +0.2083         +0.2083            26            31            26            31
 wabt/execute/taylor_pi/pi_1000000_runs_mean                      +1.8767         +1.8767        123269        354604        123261        354589
 wabt/parse/micro/eli_interpreter_mean                            -0.1446         -0.1445            23            20            23            20
 wabt/instantiate/micro/eli_interpreter_mean                      +0.2448         +0.2450            31            39            31            39
 wabt/execute/micro/eli_interpreter/halt_mean                     +1.1912         +1.1913             0             0             0             0
 wabt/execute/micro/eli_interpreter/exec105_mean                +160.7358       +160.7402             0            37             0            37
 wabt/parse/micro/factorial_mean                                  -0.0482         -0.0482             4             4             4             4
 wabt/instantiate/micro/factorial_mean                            +0.5623         +0.5624             4             6             4             6
 wabt/execute/micro/factorial/10_mean                             +2.1454         +2.1455             1             2             1             2
 wabt/execute/micro/factorial/20_mean                             +2.2646         +2.2647             1             4             1             4
 wabt/parse/micro/fibonacci_mean                                  -0.0174         -0.0174             6             5             6             5
 wabt/instantiate/micro/fibonacci_mean                            +0.5004         +0.5005             5             8             5             8
 wabt/execute/micro/fibonacci/24_mean                             +2.4647         +2.4648         10801         37422         10800         37420
 wabt/parse/micro/host_adler32_mean                               -0.2609         -0.2608             9             7             9             7
 wabt/instantiate/micro/host_adler32_mean                         +0.2724         +0.2725            13            17            13            17
 wabt/execute/micro/host_adler32/1_mean                           -0.8320         -0.8320             4             1             4             1
 wabt/execute/micro/host_adler32/100_mean                         -0.8958         -0.8958           348            36           348            36
 wabt/execute/micro/host_adler32/1000_mean                        -0.9002         -0.9002          3495           349          3494           349
 wabt/parse/micro/spinner_mean                                    -0.1024         -0.1024             4             3             4             3
 wabt/instantiate/micro/spinner_mean                              +0.6440         +0.6441             3             6             3             6
 wabt/execute/micro/spinner/1_mean                                +1.2754         +1.2755             0             0             0             0
 wabt/execute/micro/spinner/1000_mean                             +2.1001         +2.1002            25            79            25            79

```

**Comment by chfast** [comment]
I propose to merge it as is, and report upstream. Depending on the feedback we may adjust to code in following PRs.

**Comment by chfast** [comment]
Updated benchmarks, now without `-fPIC` (not needed because `wasm.so` is not built).
```
 wabt/parse/blake2b_mean                                          -0.0409         -0.0409            72            70            72            70
 wabt/instantiate/blake2b_mean                                    +0.0712         +0.0712            78            84            78            84
 wabt/execute/blake2b/512_bytes_rounds_1_mean                     +1.9298         +1.9297           154           452           154           452
 wabt/execute/blake2b/512_bytes_rounds_16_mean                    +1.9391         +1.9389          2356          6924          2356          6924
 wabt/parse/ecpairing_mean                                        -0.0030         -0.0031          4115          4103          4115          4103
 wabt/instantiate/ecpairing_mean                                  +0.0336         +0.0336          4148          4287          4147          4287
 wabt/execute/ecpairing/onepoint_mean                             +1.9303         +1.9302        720502       2111302        720507       2111217
 wabt/parse/keccak256_mean                                        -0.0083         -0.0082           120           119           120           119
 wabt/instantiate/keccak256_mean                                  +0.1017         +0.1017           123           136           123           136
 wabt/execute/keccak256/512_bytes_rounds_1_mean                   +1.5022         +1.5024           173           432           173           432
 wabt/execute/keccak256/512_bytes_rounds_16_mean                  +1.5100         +1.5101          2543          6383          2543          6384
 wabt/parse/memset_mean                                           -0.1467         -0.1466            19            16            19            16
 wabt/instantiate/memset_mean                                     +0.1527         +0.1527            23            27            23            27
 wabt/execute/memset/256_bytes_mean                               +2.1542         +2.1543            11            36            11            36
 wabt/execute/memset/60000_bytes_mean                             +2.1705         +2.1705          2521          7992          2521          7993
 wabt/parse/mul256_opt0_mean                                      -0.1455         -0.1455            24            21            24            21
 wabt/instantiate/mul256_opt0_mean                                +0.1140         +0.1140            27            30            27            30
 wabt/execute/mul256_opt0/input0_mean                             +2.4899         +2.4899            44           155            44           155
 wabt/execute/mul256_opt0/input1_mean                             +2.4910         +2.4910            45           155            45           155
 wabt/parse/ramanujan_pi_mean                                     -0.0420         -0.0420            65            62            65            62
 wabt/instantiate/ramanujan_pi_mean                               +0.0889         +0.0889            68            74            68            74
 wabt/execute/ramanujan_pi/33_runs_mean                           +1.8442         +1.8441           237           673           237           673
 wabt/parse/sha1_mean                                             -0.0231         -0.0231           113           110           113           110
 wabt/instantiate/sha1_mean                                       +0.0591         +0.0591           116           123           116           123
 wabt/execute/sha1/512_bytes_rounds_1_mean                        +2.1333         +2.1332           174           544           174           544
 wabt/execute/sha1/512_bytes_rounds_16_mean                       +2.1642         +2.1641          2415          7641          2415          7641
 wabt/parse/sha256_mean                                           -0.0237         -0.0238           198           193           198           193
 wabt/instantiate/sha256_mean                                     +0.0400         +0.0400           202           210           202           210
 wabt/execute/sha256/512_bytes_rounds_1_mean                      +1.7065         +1.7064           188           510           188           510
 wabt/execute/sha256/512_bytes_rounds_16_mean                     +1.7403         +1.7402          2599          7123          2599          7123
 wabt/parse/taylor_pi_mean                                        -0.3649         -0.3650             9             6             9             6
 wabt/instantiate/taylor_pi_mean                                  +0.0986         +0.0986            14            16            14            16
 wabt/execute/taylor_pi/pi_1000000_runs_mean                      +1.7437         +1.7436         70885        194487         70885        194481
 wabt/parse/micro/eli_interpreter_mean                            -0.2214         -0.2214            12             9            12             9
 wabt/instantiate/micro/eli_interpreter_mean                      +0.1220         +0.1220            16            18            16            18
 wabt/execute/micro/eli_interpreter/halt_mean                     +0.9112         +0.9112             0             0             0             0
 wabt/execute/micro/eli_interpreter/exec105_mean                +150.8334       +150.8263             0            19             0            19
 wabt/parse/micro/factorial_mean                                  -0.0959         -0.0959             2             2             2             2
 wabt/instantiate/micro/factorial_mean                            +0.4774         +0.4773             2             3             2             3
 wabt/execute/micro/factorial/10_mean                             +1.8356         +1.8355             0             1             0             1
 wabt/execute/micro/factorial/20_mean                             +1.9805         +1.9804             1             2             1             2
 wabt/parse/micro/fibonacci_mean                                  -0.0643         -0.0642             3             3             3             3
 wabt/instantiate/micro/fibonacci_mean                            +0.4012         +0.4013             3             4             3             4
 wabt/execute/micro/fibonacci/24_mean                             +2.1453         +2.1454          5812         18279          5812         18280
 wabt/parse/micro/host_adler32_mean                               -0.3032         -0.3032             5             3             5             3
 wabt/instantiate/micro/host_adler32_mean                         +0.1361         +0.1361             7             8             7             8
 wabt/execute/micro/host_adler32/1_mean                           -0.8747         -0.8747             2             0             2             0
 wabt/execute/micro/host_adler32/100_mean                         -0.9290         -0.9290           219            16           219            16
 wabt/execute/micro/host_adler32/1000_mean                        -0.9292         -0.9292          2184           155          2184           155
 wabt/parse/micro/spinner_mean                                    -0.1185         -0.1185             2             2             2             2
 wabt/instantiate/micro/spinner_mean                              +0.5397         +0.5397             2             3             2             3
 wabt/execute/micro/spinner/1_mean                                +0.9698         +0.9698             0             0             0             0
 wabt/execute/micro/spinner/1000_mean                             +1.7462         +1.7462            15            40            15            40
```

**Review by gumb0 — APPROVED** [review]

**Review by axic — APPROVED** [review]

**Pull request merged**